packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesController.kt

```diff
@@ -1,55 +1,83 @@
-import expo.modules.updates.loader.*
-import expo.modules.updates.loader.LoaderTask.RemoteUpdateStatus
-import expo.modules.updates.loader.LoaderTask.LoaderTaskCallback
-import expo.modules.updates.logging.UpdatesErrorCode
-import expo.modules.updates.logging.UpdatesLogReader
-import expo.modules.updates.logging.UpdatesLogger
-import expo.modules.updates.manifest.EmbeddedManifest
-import expo.modules.updates.manifest.ManifestMetadata
+import expo.modules.updates.loader.LoaderTask
 import expo.modules.updates.manifest.UpdateManifest
-import expo.modules.updates.selectionpolicy.LauncherSelectionPolicySingleUpdate
-import expo.modules.updates.selectionpolicy.ReaperSelectionPolicyDevelopmentClient
 import expo.modules.updates.selectionpolicy.SelectionPolicy
-import expo.modules.updates.selectionpolicy.SelectionPolicyFactory
-import expo.modules.updates.statemachine.UpdatesStateChangeEventSender
 import expo.modules.updates.statemachine.UpdatesStateContext
-import expo.modules.updates.statemachine.UpdatesStateEvent
-import expo.modules.updates.statemachine.UpdatesStateEventType
-import expo.modules.updates.statemachine.UpdatesStateMachine
-import expo.modules.updates.statemachine.UpdatesStateValue
-import expo.modules.updatesinterface.UpdatesInterface
-import org.json.JSONObject
 import java.io.File
-import java.lang.ref.WeakReference
 import java.util.Date
-import java.util.HashMap
+
+data class UpdatesModuleConstants(
+  val launchedUpdate: UpdateEntity?,
+  val embeddedUpdate: UpdateEntity?,
+  val isEmergencyLaunch: Boolean,
+  val isEnabled: Boolean,
+  val releaseChannel: String,
+  val isUsingEmbeddedAssets: Boolean,
+  val runtimeVersion: String?,
+  val checkOnLaunch: UpdatesConfiguration.CheckAutomaticallyConfiguration,
+  val requestHeaders: Map<String, String>,
+  val localAssetFiles: Map<AssetEntity, String>?,
+)
+
+sealed class CheckForUpdateResult(private val status: Status) {
+  private enum class Status {
+    NO_UPDATE_AVAILABLE,
+    UPDATE_AVAILABLE,
+    ROLL_BACK_TO_EMBEDDED,
+    ERROR
+  }
+
+  class NoUpdateAvailable(val reason: LoaderTask.RemoteCheckResultNotAvailableReason) : CheckForUpdateResult(Status.NO_UPDATE_AVAILABLE)
+  class UpdateAvailable(val updateManifest: UpdateManifest) : CheckForUpdateResult(Status.UPDATE_AVAILABLE)
+  class RollBackToEmbedded(val commitTime: Date) : CheckForUpdateResult(Status.ROLL_BACK_TO_EMBEDDED)
+  class ErrorResult(val error: Exception, val message: String) : CheckForUpdateResult(Status.ERROR)
+}
+
+sealed class FetchUpdateResult(private val status: Status) {
+  private enum class Status {
+    SUCCESS,
+    FAILURE,
+    ROLL_BACK_TO_EMBEDDED,
+    ERROR
+  }
+
+  class Success(val update: UpdateEntity) : FetchUpdateResult(Status.SUCCESS)
+  class Failure : FetchUpdateResult(Status.FAILURE)
+  class RollBackToEmbedded : FetchUpdateResult(Status.ROLL_BACK_TO_EMBEDDED)
+  class ErrorResult(val error: Exception) : FetchUpdateResult(Status.ERROR)
+}
+
+interface IUpdatesController {
```

Comment:
split this into two: public API (just start I think), internal API. make internal extend public